### ISSUE TITLE
Bug is back in 2.14.0  -- UniqueConstraintViolationException and subsequent campaign failure

### ISSUE THREAD

**Issue by namankumar** [mautic/mautic#6613]
**Please DO NOT report security vulnerabilities here. Send them to security@mautic.com instead.**

| Q  | A
| ---| ---
| Bug report? | x |
| Feature request? || 
| Enhancement? | |

## Description:

I'm facing the issue described in #3749

Steps to reproduce and Mautic behaviour are exactly as described. The difference in logs is that my logs show that the Entity Manager is closed. See below.

This bug goes away when I downgrade to 2.13.1. (would not like to keep using 2.13.1 due to stability issues but have to while this bug exists)

## If a bug:

| Q   | A
| --- | ---
| Mautic version | 2.14.0
| PHP version | 7.0

### Steps to reproduce:
See #3749 
 
### Log errors: 
```
[2018-09-12 22:37:03] mautic.ERROR: CAMPAIGN: The EntityManager is closed. [] []

[2018-09-12 22:38:02] mautic.ERROR: CAMPAIGN: An exception occurred while executing ‘INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)’ with params [1, “2018-09-12 22:38:02”, 0, null, 1, “a:0:{}“, “email”, 101, 0, 37, 37510, 9, 2999]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry ‘37-37510-1’ for key ‘campaign_rotation’ [] []

[2018-09-12 22:38:02] mautic.ERROR: CAMPAIGN: The EntityManager is closed. [] []

```
_Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable._

**Comment by demondehellis** [comment]
The same in 2.14.1
```
[2018-09-30 08:11:54] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2018-09-30 08:11:54", 0, null, 1, "a:0:{}", null, null, 0, 2571, 246371, 76, 5013130]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '2571-246371-1' for key 'campaign_rotation' [] []
[2018-09-30 08:11:54] mautic.ERROR: CAMPAIGN: The EntityManager is closed. [] []
```

**Comment by heathdutton** [comment]
Are you using the `--force` flag perchance, or do you have any other errors in your PHP system error_log?

I ask because this can be a catch all error for other things. Not something I'm able to reproduce with 2.14.x

**Comment by Shunshifu** [comment]
This is the same install that @demondehellis mentioned.

These are the errors I'm seeing now

[2018-10-01 16:48:55] mautic.NOTICE: PHP Notice - Undefined index: status - in file /var/www/mautic/app/bundles/NotificationBundle/Views/SubscribedEvents/Timeline/index.html.php - at line 19 [] []
[2018-10-01 16:48:55] mautic.NOTICE: PHP Notice - Undefined index: type - in file /var/www/mautic/app/bundles/NotificationBundle/Views/SubscribedEvents/Timeline/index.html.php - at line 21 [] []
[2018-10-01 16:48:55] mautic.NOTICE: PHP Notice - Undefined index: heading - in file /var/www/mautic/app/bundles/NotificationBundle/Views/SubscribedEvents/Timeline/index.html.php - at line 25 [] []
[2018-10-01 16:48:55] mautic.NOTICE: PHP Notice - Undefined index: content - in file /var/www/mautic/app/bundles/NotificationBundle/Views/SubscribedEvents/Timeline/index.html.php - at line 27 [] []

**Comment by Shunshifu** [comment]
This bug is crushing us. The system runs in fits and starts. Some campaigns run, others not at all. @demondehellis has got us limping along by batching to stop the memory leak. I'm seeing the error popup every few hours.

Unfortunately for us, we worked all summer to be ready for our busy season (which is now) and we are now mostly dead in the water.
What other information can we provide to help get this fixed?
Thanks In Advance

**Comment by Shunshifu** [comment]
I think that removing "jump to Event" from the campaign stopped this. I will do more testing and watching of logs to verify

**Comment by heathdutton** [comment]
Well, more logs would help. You could also tun  `php app/console doctrine:schema:update --dump-sql' to make sure there aren't some migrations that got missed.

**Comment by Shunshifu** [comment]
OK thanks. I'll have @demondehellis get together as much info as possible and follow your suggestion.

**Comment by namankumar** [comment]
@Shunshifu yes, removing two types of actions I think resolve this issue: 1. jumping to event within the campaign 2. Firing an event (send email) after a certain time period (7 days). This hack can get you started it again but it requires a big change in your campaign design. For 1, you can make multiple campaigns and instead of "jump to event" you can do "move to campaign". For 2, I am not sure...

**Comment by Shunshifu** [comment]
@namankumar  Thank you.  
1 I will do on my main campaign that's giving me trouble.

2 I don't have any send email after x days but I do have check condition after x days which leads to send email if true. Hopefully that doesn't cause an issue

Thanks for the response

**Comment by demondehellis** [comment]
I can say that our sending problem was not related to this issue. It was related to sending a big number of marketing/segment emails and after playing with params it seems to start sending. 

Currently we are using

```
mautic/app/console mautic:emails:send --message-limit=1200 --time-limit=50 --do-not-clear
```

We are using "--do-not-clear" because in other case it works too slow or just stuck when we have about 80k files in our "app/spool" dir. I understand what it does, but it have to be reworked.

Without $skipClear it just falls while reading 100k+ files
```
ProcessEmailQueueCommand.php:90
$finder = Finder::create()->in($spoolPath)->name('*.{finalretry,sending,tryagain}');
```

Also, I think 'jumps' are working fine, we performed some tests with simple campaigns.

I no longer see errors I posted previously in our logs, I'm not sure exactly why they disappeared. We did a lot of changes and tried many thing related to cron. But we didn't use --force mode. We have other errors related to lead_event_log but I think they point to something different ( to a "page hit" event ).

**Comment by namankumar** [comment]
This error is being better handled in 2.14.2. 

I've successfully executed over > 1000 Jump to Event in the last 24 hours. I still get the error sometimes but the EntityManager does not close thus allowing the execution to continue. Will continue to test over the next few days and report back.

This error is
`Duplicate entry ‘37-37510-1’ for key ‘campaign_rotation’`

**Comment by carloshmm** [comment]
I end up with the same problem:
![mautic](https://user-images.githubusercontent.com/31554436/51058632-8fb69800-15d0-11e9-8922-319b271ab093.PNG)

The "Jump to Event" on my campaign called: "01 Timer checagem email" is redirecting the campaign to the event "01 Verifica email 01", every time the cronjob executes the "mautic:campaigns:trigger" event I'm getting these error messages:

```
[2019-01-11 20:25:43] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2019-01-11 20:25:43", 0, null, 1, "a:0:{}", null, null, 0, 999, 959, 25, 20]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '999-959-1' for key 'campaign_rotation' [] []
[2019-01-11 20:28:31] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2019-01-11 20:28:31", 0, null, 1, "a:0:{}", null, null, 0, 999, 959, 25, 20]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '999-959-1' for key 'campaign_rotation' [] []
[2019-01-11 20:28:33] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2019-01-11 20:28:33", 0, null, 1, "a:0:{}", null, null, 0, 999, 959, 25, 20]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '999-959-1' for key 'campaign_rotation' [] []
[2019-01-11 20:32:43] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2019-01-11 20:32:43", 0, null, 1, "a:0:{}", null, null, 0, 999, 959, 25, 20]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '999-959-1' for key 'campaign_rotation' [] []
```

What I think interesting is that the example showed on the official website most likely will not work as well since it is returning to an event that was already executed on the campaign flow:
https://mautic.com/help/august-2018-update/#jump-to-event

![](https://mautic.com/wp-content/uploads/2018/08/Screen-Shot-2018-08-06-at-5.12.04-PM-768x438.png)

As @namankumar said, I think this is related with returning to an past event already executed by that lead on the campaign.

**Comment by dsp76** [comment]
Looks like this is also happening with 2.15.0?

**Comment by MauriceW2** [comment]
Yes, the bug is still in .15.0.

I just discovered it, because some campaigns are not running.

It happens to fresh contacts, even in super simple straightforward campaigns, with just setting a tag and sending a notification to the contact owner.

[2019-03-12 13:37:03] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO mautic_campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2019-03-12 13:37:03", 0, null, 1, "a:0:{}", "email", 18, 0, 249, 2079, 7, 2]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '249-2079-1' for key 'mautic_campaign_rotation' [] []
[2019-03-12 13:37:03] mautic.ERROR: CAMPAIGN: The EntityManager is closed. [] []

**Comment by MauriceW2** [comment]
I fixed it, at least temporarily. 

**What did I do:**
It says: Duplicate entry '249-2079-1' for key 'mautic_campaign_rotation'. This could mean that a value inserted into the DB must be unique, while it's not.

I went into phpMyAdmin and opened the table maut_campaign_lead_event_leg. Tjis is where the error occurs.

I then clicked 'Structure', scrolled down (below the columns) and clicked the + sign to see the indexes.

Then I clicked on  Edit the index 'maut_campaign_rotation'. This is the key. I changed the Indextype from UNIQUE to INDEX.

Clicked start to save the index.

Flushed my log and now it works. The error is gone and my campaigns are running again.

Still I need to investigate more if this change has more (unwanted) impact.

**Comment by Uebix** [comment]
Hello, this error persist in 2.15.1
mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO....'
SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '140-207837-2681' for key 'phplist_campaign_rotation'

To test the problem i create a very simple campaign with a test segment.
The last "Jump to event" is the problem.
![image](https://user-images.githubusercontent.com/3381177/55283596-217ba900-5367-11e9-8dc5-e0cb4ad598a5.png)

**Comment by pierrederuel** [comment]
Any news on that problem? @MauriceW2 did you notice side effects with the update in the table structure? Thanks

**Comment by MauriceW2** [comment]
Hi @pierrederuel,

Nope, nothing heard about it anymore. Also no side effects after using this mod for almost a year. All works fine and stable.

**Comment by pierrederuel** [comment]
Great! @MauriceW2  Thanks for your feedback!

**Comment by regevbr** [comment]
I'm facing the same issue. I will apply the temp fix and see if it helps.

**Comment by MauriceW2** [comment]
@regevbr, please let us know if it works for you.

**Comment by regevbr** [comment]
sure thing

**Comment by dichvuhuuich** [comment]
> I fixed it, at least temporarily.
> 
> **What did I do:**
> It says: Duplicate entry '249-2079-1' for key 'mautic_campaign_rotation'. This could mean that a value inserted into the DB must be unique, while it's not.
> 
> 1. I went into phpMyAdmin and opened the table maut_campaign_lead_event_log. This is where the error occurs.
> 2. I then clicked 'Structure', scrolled down (below the columns) and clicked the + sign to see the indexes.
> 3. Then I clicked on link to edit the index 'maut_campaign_rotation'. This is the key.
> 4. I changed the Indextype from UNIQUE to INDEX.
> 5. Clicked start to save the index.
> 
> I flushed my log and now it works. The error is gone and my campaigns are running again.
> 
> I still need to investigate more if this change has more (unwanted) impact.
> 
> Any feedback would be great.

I'm using 2.15.1, all works normally until yesterday, this error happens, I follow your instructions to fix, now my campaigns work normally again. I'll watch out if this change has more (unwanted) impact I'll feed back.

**Comment by regevbr** [comment]
The fix is working well for me for a long time now!

**Comment by MauriceW2** [comment]
After using a fresh Mautic install for 3 months, I suddenly ran into this same issue.

After re-applying the fix I suggested, the bug was solved.

@escopecz, you added the bug label on 18 Oct 2018. 

Can it be implemented for testing to get this in a next update?

**Comment by npracht** [comment]
@escopecz can we make a PR out of that ---> https://github.com/mautic/mautic/issues/6613#issuecomment-472052071

**Comment by escopecz** [comment]
@alanhartless as the author of the rotation do you remember if the `maut_campaign_rotation` index must be unique?

**Comment by alanhartless** [comment]
The bug is not the unique index as that is there for a reason namely the code expects there to be only a single entry per event, contact and rotation. If there's a bug, it's around why the rotation is not getting incremented each time the event is expected to be executed again.

**Comment by pety-dc** [comment]
I don't know the code for campaign restart feature. but I think it could be possible that if you use `jump` event to jump backwards in the campaign rather than forward or to another branch, the rotation doesn't get increased, does it?

**Comment by pety-dc** [comment]
I tried to debug this.

I believe the the jump only fails is the jump campaign event is **scheduled**

`Mautic\CampaignBundle\Executioner\EventExecutioner` has a method `executeEventsForContacts` which contains a part where it isolates the jump and non-jump campaign events and first executes the non-jump events then increments the `rotation` via `fetchRotationAndGenerateLogsFromContacts` and only then executes the jump events.

```
public function executeEventsForContacts(ArrayCollection $events, ArrayCollection $contacts, Counter $childrenCounter = null, $isInactive = false)
    {
        if (!$contacts->count()) {
            return;
        }

        // Schedule then return those that need to be immediately executed
        $executeThese = $this->scheduleEvents($events, $contacts, $childrenCounter, $isInactive);

        // Execute non jump-to events normally
        $otherEvents = $executeThese->filter(function (Event $event) {
            return CampaignActionJumpToEventSubscriber::EVENT_NAME !== $event->getType();
        });

        if ($otherEvents->count()) {
            foreach ($otherEvents as $event) {
                $this->executeForContacts($event, $contacts, $childrenCounter, $isInactive);
            }
        }

        // Now execute jump to events
        $jumpEvents = $executeThese->filter(function (Event $event) {
            return CampaignActionJumpToEventSubscriber::EVENT_NAME === $event->getType();
        });
        if ($jumpEvents->count()) {
            $jumpLogs = [];

            // Create logs for the jump to events before the rotation is incremented
            foreach ($jumpEvents as $key => $event) {
                $config         = $this->collector->getEventConfig($event);
                $jumpLogs[$key] = $this->eventLogger->fetchRotationAndGenerateLogsFromContacts($event, $config, $contacts, $isInactive);
            }

            // Increment the campaign rotation for the given contacts and current campaign
            $this->leadRepository->incrementCampaignRotationForContacts(
                $contacts->getKeys(),
                $jumpEvents->first()->getCampaign()->getId()
            );

            // Process the jump to events
            foreach ($jumpLogs as $key => $logs) {
                $this->executeLogs($jumpEvents->get($key), $logs, $childrenCounter);
            }
        }
    }
```

`Mautic\CampaignBundle\Executioner\KickoffExecutioner` and `Mautic\CampaignBundle\Executioner\RealTimeExecutioner` executes events via `$this->executioner->executeEventsForContact`

But `Mautic\CampaignBundle\Executioner\ScheduledExecutioner` calls `$this->executioner->executeLogs` method directly

```
private function executeScheduled($eventId, \DateTime $now)
    {
        $logs = $this->repo->getScheduled($eventId, $this->now, $this->limiter);
        while ($logs->count()) {
            try {
                $this->scheduledContactFinder->hydrateContacts($logs);
            } catch (NoContactsFoundException $e) {
                break;
            }

            $event = $logs->first()->getEvent();
            $this->progressBar->advance($logs->count());
            $this->counter->advanceEvaluated($logs->count());

            // Validate that the schedule is still appropriate
            $this->validateSchedule($logs, $now);

            // Execute if there are any that did not get rescheduled
            $this->executioner->executeLogs($event, $logs, $this->counter);

            // Get next batch
            $this->scheduledContactFinder->clear();
            $logs = $this->repo->getScheduled($eventId, $this->now, $this->limiter);
        }
    }
```

Then the CampaignActionJumpToEventSubscriber::onJumpToEvent method handles the jump but the following steps lead to failure due to the missed rotation increase 

```
Exception trace:
 () at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:105
 PDOStatement->execute() at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:105
 Doctrine\DBAL\Driver\PDOStatement->execute() at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Statement.php:168
 Doctrine\DBAL\Statement->execute() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:281
 Doctrine\ORM\Persisters\Entity\BasicEntityPersister->executeInserts() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:1014
 Doctrine\ORM\UnitOfWork->executeInserts() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:378
 Doctrine\ORM\UnitOfWork->commit() at /var/www/html/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:356
 Doctrine\ORM\EntityManager->flush() at /var/www/html/app/bundles/CoreBundle/Entity/CommonRepository.php:785
 Mautic\CoreBundle\Entity\CommonRepository->saveEntities() at /var/www/html/app/bundles/CampaignBundle/Executioner/Logger/EventLogger.php:265
 Mautic\CampaignBundle\Executioner\Logger\EventLogger->persistPendingAndInsertIntoLogStack() at /var/www/html/app/bundles/CampaignBundle/Executioner/Logger/EventLogger.php:151
 Mautic\CampaignBundle\Executioner\Logger\EventLogger->persistQueuedLogs() at /var/www/html/app/bundles/CampaignBundle/Executioner/Logger/EventLogger.php:247
 Mautic\CampaignBundle\Executioner\Logger\EventLogger->generateLogsFromContacts() at /var/www/html/app/bundles/CampaignBundle/Executioner/Logger/EventLogger.php:217
 Mautic\CampaignBundle\Executioner\Logger\EventLogger->fetchRotationAndGenerateLogsFromContacts() at /var/www/html/app/bundles/CampaignBundle/Executioner/EventExecutioner.php:193
 Mautic\CampaignBundle\Executioner\EventExecutioner->executeForContacts() at /var/www/html/app/bundles/CampaignBundle/EventListener/CampaignActionJumpToEventSubscriber.php:108
 Mautic\CampaignBundle\EventListener\CampaignActionJumpToEventSubscriber->onJumpToEvent() at /var/www/html/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:61
 Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke() at /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:184
 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() at /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:46
 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() at /var/www/html/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:133
 Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch() at /var/www/html/app/bundles/CampaignBundle/Executioner/Dispatcher/ActionDispatcher.php:100
 Mautic\CampaignBundle\Executioner\Dispatcher\ActionDispatcher->dispatchEvent() at /var/www/html/app/bundles/CampaignBundle/Executioner/Event/ActionExecutioner.php:73
 Mautic\CampaignBundle\Executioner\Event\ActionExecutioner->execute() at /var/www/html/app/bundles/CampaignBundle/Executioner/EventExecutioner.php:228
 Mautic\CampaignBundle\Executioner\EventExecutioner->executeLogs() at /var/www/html/app/bundles/CampaignBundle/Executioner/ScheduledExecutioner.php:323
 Mautic\CampaignBundle\Executioner\ScheduledExecutioner->executeScheduled() at /var/www/html/app/bundles/CampaignBundle/Executioner/ScheduledExecutioner.php:291
 Mautic\CampaignBundle\Executioner\ScheduledExecutioner->executeOrRecheduleEvent() at /var/www/html/app/bundles/CampaignBundle/Executioner/ScheduledExecutioner.php:150
 Mautic\CampaignBundle\Executioner\ScheduledExecutioner->execute() at /var/www/html/app/bundles/CampaignBundle/Command/TriggerCampaignCommand.php:417
 Mautic\CampaignBundle\Command\TriggerCampaignCommand->executeScheduled() at /var/www/html/app/bundles/CampaignBundle/Command/TriggerCampaignCommand.php:365
 Mautic\CampaignBundle\Command\TriggerCampaignCommand->triggerCampaign() at /var/www/html/app/bundles/CampaignBundle/Command/TriggerCampaignCommand.php:285
 Mautic\CampaignBundle\Command\TriggerCampaignCommand->execute() at /var/www/html/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/console/Application.php:853
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/symfony/framework-bundle/Console/Application.php:84
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/html/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at /var/www/html/app/console:41
```

@alanhartless What do you think?

**Comment by pety-dc** [comment]
I tested a countermeasure:
If another type of action preceeds the jumpTo, then the jumpTo runs smoothly with increased rotation.

So if you have a scheduled jumpTo event you should rewrite it to a scheduled Event of any type which is followed by the jumpTo immediately.

**Comment by zaharovrd** [comment]
i deleted all jumpTo event from campaign, do campaign copy, but it didn't work :(

**Comment by zaharovrd** [comment]
> I tested a countermeasure:
> If another type of action preceeds the jumpTo, then the jumpTo runs smoothly with increased rotation.
> 
> So if you have a scheduled jumpTo event you should rewrite it to a scheduled Event of any type which is followed by the jumpTo immediately.

i deleted all jumpTo event from campaign, do campaign copy, but it didn't work :(

**Comment by pety-dc** [comment]
@zaharovrd 
Sorry for the dumb questions, but if you duplicated the campaign with no jumpTo events, then are you sure that it was the duplicated event that failed and not the original? I mean did you do your test with the original campaign disabled?

**Comment by zaharovrd** [comment]
> @zaharovrd
> Sorry for the dumb questions, but if you duplicated the campaign with no jumpTo events, then are you sure that it was the duplicated event that failed and not the original? I mean did you do your test with the original campaign disabled?
Yes, i disabled original campaign and tested on copy

in logs the mautic is issued

> [2019-11-12 12:44:04] mautic.ERROR: CAMPAIGN: An exception occurred while executing 'INSERT INTO campaign_lead_event_log (rotation, date_triggered, is_scheduled, trigger_date, system_triggered, metadata, channel, channel_id, non_action_path_taken, event_id, lead_id, campaign_id, ip_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [1, "2019-11-12 12:44:04", 0, null, 1, "a:0:{}", "email", 19, 0, 64, 39858, 6, 16420]:  SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '64-39858-1' for key 'campaign_rotation' [] []
> [2019-11-12 12:44:04] mautic.ERROR: CAMPAIGN: The EntityManager is closed. [] []
> [2019-11-12 12:44:04] mautic.ERROR: CAMPAIGN: The EntityManager is closed. [] []
>

**Comment by TnowL** [comment]
Any update on this? I also have this same issue on version 2.16.2. 

I've tried @MauriceW2 fix and it works but there is a few minor side effects. 

> 
> 
> I fixed it, at least temporarily.
> 
> **What did I do:**
> It says: Duplicate entry '249-2079-1' for key 'mautic_campaign_rotation'. This could mean that a value inserted into the DB must be unique, while it's not.
> 
>     1. I went into phpMyAdmin and opened the table maut_campaign_lead_event_log. This is where the error occurs.
> 
>     2. I then clicked 'Structure', scrolled down (below the columns) and clicked the + sign to see the indexes.
> 
>     3. Then I clicked on link to edit the index 'maut_campaign_rotation'. This is the key.
> 
>     4. I changed the Indextype from UNIQUE to INDEX.
> 
>     5. Clicked start to save the index.
> 
> 
> I flushed my log and now it works. The error is gone and my campaigns are running again.
> 
> I still need to investigate more if this change has more (unwanted) impact.
> 
> Any feedback would be great.


Under campaign > actions you will see success rate and pending are miscalculated. And because of that miscalculation of the success rate the progress bar (green bar) can grow of the screen according to the percent of success rate. Example 1000% will have a width of 1000%.
![jump-to-events](https://user-images.githubusercontent.com/47170989/81637455-995c4400-9440-11ea-9581-d980bee0abc7.jpg)

**Comment by pawonpawon** [comment]
The bug is still there - wasted DAYS on this. Was ready to remove Mautic. 

Actions were randomly not performed. First, though I did something wrong because I am new to Mautic.

Then tested on a new Campaign with only 1 action - nothing happens.

Yes, I too have 'jump' actions that jump back above in the same campaign.

And yes, this happens in ANY campaign afterward, when a contact went through the 'problem' campaign.

For now I used the same temporary fix. I used this SQL:

```
ALTER TABLE `database-name`.campaign_lead_event_log DROP INDEX campaign_rotation;

CREATE INDEX campaign_rotation USING BTREE ON `database-name`.campaign_lead_event_log (event_id,lead_id,rotation);
```

All is working fine now - finally I see things happening that are supposed to be happening (like sending emails, updating tags)

Thanks @MauriceW2 

And yes, I think you are right @pety-dc - that must be the source of the problem, but it's late and I don't have time right now to deep-dive into it.

**Comment by alexhammerschmied** [comment]
I have the same problem with 2.16.2 
The jump events which are scheduled to a certain time don't get executed. They only get scheduled but then nothing happens. In the logs is the same error. 

I changed the campaign the following:
- Instead of scheduling a jump after x minutes, 
+ Add a tag after x minutes
+ make the jump immediately after the tag 

I'll let you know in a few ours how it works. 
EDIT: It works perfectly. :)

**Comment by ericross** [comment]
I was also getting the duplicate key error with 2.16.2 and the temporary fix suggested that drops the unique index has allowed me to continue.

**Comment by alexhammerschmied** [comment]
As described earlier, I tested it with creating a tag action before the jump event and it works fine. 
So instead of messing around with the database, I suggest to just make this extra step instead.

**Comment by TnowL** [comment]
> 
> 
> I have the same problem with 2.16.2
> The jump events which are scheduled to a certain time don't get executed. They only get scheduled but then nothing happens. In the logs is the same error.
> 
> I changed the campaign the following:
> 
>     * Instead of scheduling a jump after x minutes,
> 
> 
>     * Add a tag after x minutes
> 
>     * make the jump immediately after the tag
> 
> 
> I'll let you know in a few ours how it works.
> EDIT: It works perfectly. :)

Confirmed @alexhammerschmied workaround works without messing with the database.

**Comment by zaharovrd** [comment]
`* Add a tag after x minutes`
which tag?
maybe screenshot?

**Comment by TnowL** [comment]
> 
> 
> `* Add a tag after x minutes`
> which tag?
> maybe screenshot?

Action > Modify contact's tags

**Comment by zaharovrd** [comment]
What did you name the tag?
For example, "does not read letters"?

**Comment by alexhammerschmied** [comment]
@zaharovrd yes for example. It doesn't really matter. It is only for you and the purpose of the workaround. You could also name it a1, a2, b1, b2, c1,...

**Comment by zaharovrd** [comment]
I got a bit, and if I have email 1 and the condition for opening it:
"if you read email 1, then send email 2.
"if you have not read email 1, then send email 1_1"

If, under the positive condition of opening email 1_1, I just send email 2, will the chain go no further (email 3, email 4...)?

**Comment by dominicplum** [comment]
@alexhammerschmied but you update campaign or do clone campaign and repair by adding tag before? Because I update campaign and this not working for me still. Also I clone and repair by adding tag before jump to event link to form and still not working.

**Comment by kuzmany** [comment]
Fix scheduled jump to event campaign action https://github.com/mautic/mautic/pull/7129

**Comment by RCheesley** [comment]
@namankumar @alexhammerschmied @ericross @TnowL @zaharovrd @dongamini92 can you confirm whether this has been resolved with #7129 or is this still a problem? The PR will ship with the next patch release at the end of December but you can manually apply it.
